code/modules/station_goals/bluespace_tap.dm

```diff
@@ -0,0 +1,380 @@
+
+//power use and point generation
+/obj/machinery/power/bluespace_tap/process()
+	actual_power_usage = get_power_use(input_level)
+	if(surplus() < actual_power_usage)	//not enough power, so turn down a level
+		input_level--
+		return	// and no mining gets done
+	else
+		if(actual_power_usage)
+			add_load(actual_power_usage)
+			var/points_to_add = (input_level + emagged) * base_points
+			points += points_to_add	//point generation, emagging gets you 'free' points at the cost of higher anomaly chance
+			total_points += points_to_add
+		// actual input level changes slowly
+		if(input_level < desired_level && (surplus() >= get_power_use(input_level + 1)))
+			input_level++
+		else if(input_level > desired_level)
+			input_level--
+		if(prob(input_level - safe_levels + (emagged * 5)))	//at dangerous levels, start doing freaky shit. prob with values less than 0 treat it as 0
+			GLOB.event_announcement.Announce("Unexpected power spike during Bluespace Harvester Operation. Extra-dimensional intruder alert. Expected location: [get_area(src).name]. [emagged ? "DANGER: Emergency shutdown failed! Please proceed with manual shutdown." : "Emergency shutdown initiated."]", "Bluespace Harvester Malfunction")
+			if(!emagged)
+				input_level = 0	//emergency shutdown unless we're sabotaged
+				desired_level = 0
+			for(var/i = 1, i <= rand(1, 3), i++)
+				var/turf/location = locate(x + rand(-5,5), y + rand(-5,5), z)
+				new /obj/structure/spawner/nether/bluespace_tap(location)
+
+
+
+/obj/machinery/power/bluespace_tap/tgui_data(mob/user)
+	var/data[0]
+
+	data["desiredLevel"] = desired_level
+	data["inputLevel"] = input_level
+	data["points"] = points
+	data["totalPoints"] = total_points
+	data["powerUse"] = actual_power_usage
+	data["availablePower"] = surplus()
+	data["maxLevel"] = max_level
+	data["emagged"] = emagged
+	data["safeLevels"] = safe_levels
+	data["nextLevelPower"] = get_power_use(input_level + 1)
+
+	var/list/listed_items = list()//a list of lists, each inner list equals a datum
+	for(var/key = 1 to length(product_list))
+		var/datum/data/bluespace_tap_product/A = product_list[key]
+		listed_items.Add(list(list(	//note that adding a list to a list in BYOND adds the indiviudal list elements
+				"key" = key,		//so to get a list of lists the inner lists need to be wrapped in another list
+				"name" = A.product_name,
+				"price" = A.product_cost)))
+	data["product"] = listed_items
+	return data
+
+
+/obj/machinery/power/bluespace_tap/attack_hand(mob/user)
+	add_fingerprint(user)
+	tgui_interact(user)
+
+/obj/machinery/power/bluespace_tap/attack_ghost(mob/user)
+	tgui_interact(user)
+
+/obj/machinery/power/bluespace_tap/attack_ai(mob/user)
+	tgui_interact(user)
+
+//produces and vends the product with the desired key
+/obj/machinery/power/bluespace_tap/proc/produce(key)
+	var/datum/data/bluespace_tap_product/A = product_list[key]
```

Comment:
need to check here if the key falls within the allowed length since it is input from an external source